File: 307_NumArray/NumArray.go
/*
 * @Date: 2022-04-03 23:23:18
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-04-03 23:38:00
 * @FilePath: /algorithm/307_NumArray/NumArray.go
 */

package main

type NumArray struct {
	nums, tree []int
}

func Constructor(nums []int) NumArray {
	tree := make([]int, len(nums)+1)
	na := NumArray{nums, tree}
	for i, num := range nums {
		na.add(i+1, num)
	}
	return na
}

func (na *NumArray) add(index, val int) {
	for ; index < len(na.tree); index += index & -index {
		na.tree[index] += val
	}
}

func (na *NumArray) prefixSum(index int) (sum int) {
	for ; index > 0; index &= index - 1 {
		sum += na.tree[index]
	}
	return
}

func (na *NumArray) Update(index, val int) {
	na.add(index+1, val-na.nums[index])
	na.nums[index] = val
}

func (na *NumArray) SumRange(left, right int) int {
	return na.prefixSum(right+1) - na.prefixSum(left)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	n := Constructor([]int{1, 3, 5})
	assert(n.SumRange(0, 2) == 9)
	n.Update(1, 2)
	assert(n.SumRange(0, 2) == 8)
}

File: rust/303_NumArray/NumArray.rs
/*
 * @Date: 2024-03-18
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-03-18
 * @FilePath: /algorithm/rust/303_NumArray/NumArray.rs
 */

struct NumArray {
    s: Vec<i32>,
}

impl NumArray {
    fn new(nums: Vec<i32>) -> Self {
        let mut s = vec![0; nums.len() + 1];
        for (i, &x) in nums.iter().enumerate() {
            s[i + 1] = s[i] + x;
        }
        Self { s }
    }

    fn sum_range(&self, left: i32, right: i32) -> i32 {
        self.s[right as usize + 1] - self.s[left as usize]
    }
}

fn main() {
    let num_array = NumArray::new(vec![-2, 0, 3, -5, 2, -1]);
    assert_eq!(num_array.sum_range(0, 2), 1);
    assert_eq!(num_array.sum_range(2, 5), -1);
    assert_eq!(num_array.sum_range(0, 5), -3);
}
